aloha-core/src/main/scala/com/eharmony/aloha/dataset/StatefulRowCreator.scala

```diff
@@ -0,0 +1,123 @@
+
+  /**
+    * Some initial state that can be used on the very first call to `apply(A, S)`.
+    * @return some state.
+    */
+  val initialState: S
+
+  /**
+    * Given an `a` and some `state`, produce output, including a new state.
+    *
+    * When using this function, the user is responsible for keeping track of,
+    * and providing the state.
+    *
+    * The implementation of this function should be referentially transparent.
+    *
+    * @param a input
+    * @param state the state
+    * @return a tuple where the first element is a Tuple2 whose first element is
+    *         missing and error information and second element is an optional result.
+    *         The second element of the outer Tuple2 is the new state.
+    */
+  def apply(a: A, state: S): ((MissingAndErroneousFeatureInfo, Option[B]), S)
+
+  /**
+    * Apply the `apply(A, S)` method to the elements of the iterator.  In the first
+    * application of `apply(A, S)`, `state` will be used as the state.  In subsequent
+    * applications, the state will come from the state generated in the output of the
+    * previous application of `apply(A, S)`.
+    *
+    * ''This variant of mapping with state is'' '''non-strict''', so if that's a requirement,
+    * prefer this function over the `mapSeqWithState` variant.  Note that for this method
+    * to work, the first element is computed eagerly.
+    *
+    * To verify non-strictness, this method could be rewritten as:
+    *
+    * {{{
+    * def statefulMap[A, B, S](as: Iterator[A], s: S)
+    *                         (f: (A, S) => (B, S)): Iterator[(B, S)] = {
+    *   if (as.isEmpty)
+    *     Iterator.empty
+    *   else {
+    *     val firstA = as.next()
+    *     val initEl = f(firstA, s)
+    *     as.scanLeft(initEl){ case ((_, newS), a) => f(a, newS) }
+    *   }
+    * }
+    * }}}
+    *
+    * Then using the method, it's easy to verify non-strictness:
+    *
+    * {{{
+    * // cycleModulo4 and res are infinite.
+    * val cycleModulo4 = Iterator.iterate(0)(s => (s + 1) % 4)
+    * val res = statefulMap(cycleModulo4, 0)((a, s) => ((a + s).toDouble, s + 1))
+    *
+    * // returns:
+    * res.take(8)
+    *    .map(_._1)
+    *    .toVector
+    *    .groupBy(identity)
+    *    .mapValues(_.size)
+    *    .toVector
+    *    .sorted
+    *    .foreach(println)
+    *
+    * res.size // never returns
+    * }}}
+    *
+    * @param as Note the first element of `as` ''will be forced'' in this method in order
+    *           to construct the output.
+    * @param state the initial state to use at the start of the iterator.
+    * @return an iterator containing the `a` mapped to a
+    *         `(MissingAndErroneousFeatureInfo, Option[B])` along with the resulting
+    *         state that is created in the process.
+    */
+  def mapIteratorWithState(as: Iterator[A], state: S): Iterator[((MissingAndErroneousFeatureInfo, Option[B]), S)] = {
```

Comment:
@jmorra, do you think I should extract this to an object:

```scala
package com.eharmony.aloha.util

object MappingUtils {
  def mapIteratorWithState[A, B, S](as: Iterator[A], state: S)(f: (A, S) => (B, S)): Iterator[(B, S)] = {
    // implementation filled in
  }

  def mapSeqWithState[A, B, S](as: Seq[A], state: S)(f: (A, S) => (B, S)): (Vector[B], S) = {
    // implementation filled in
  }
}
```

And then have a trait:

```scala
trait MapWithState[A, B, S] { self: ((A, S) => (B, S)) => 
  def mapIteratorWithState(as: Iterator[A], state: S): Iterator[(B, S)] = 
    MappingUtils.mapIteratorWithState(as, state)(apply)

  def mapSeqWithState(as: Seq[A], state: S): (Vector[B], S) = 
    MappingUtils.mapSeqWithState(as, state)(apply)
}
```

Reply:
I don't understand why you don't model both of these functions together with a common interface such as `TraversableOnce`.

Reply:
@jmorra what do you think about something like this:

```scala
import scala.collection.SeqLike
import scala.collection.generic.CanBuildFrom
import scala.collection.breakOut
import scala.collection.{immutable => sci}

def statefulMap[A, B, S, Repr <: sci.Seq[A], That]
    (as: SeqLike[A, Repr], s: S)
    (f: (A, S) => (B, S))
    (implicit cbf: CanBuildFrom[Repr, (B, S), That]): That = {
  if (as.isEmpty)
    cbf().result()
  else {
    val initEl = f(as.head, s)
    as.tail.scanLeft(initEl){ case ((_, newS), a) => f(a, newS) }(breakOut)
  }
}
```

Then we can use any immutable sequence like you suggested and we can maintain the sequence's structure by not supplying a `scala.collection.generic.CanBuildFrom` or we can change the structure by supplying a `CanBuildFrom`.  See below:

```scala
val sCycleModulo4 = Stream.iterate(0)(s => (s + 1) % 4)
val vCycleModulo4 = Vector.iterate(0, 10)(s => (s + 1) % 4)
val addAndIncrement = (a: Int, s: Int) => ((a + s).toDouble, s + 1)

val stream = statefulMap(sCycleModulo4, 0)(addAndIncrement)

val list: List[(Double, Int)] = 
  statefulMap(vCycleModulo4, 0)(addAndIncrement)(breakOut)
```

We also get safety guarantees for things that are not immutable (e.g. `Array`).  For instance:

```scala
val aCycleModulo4 = Array.iterate(0, 10)(s => (s + 1) % 4)
statefulMap(aCycleModulo4, 0)(addAndIncrement)
```

Returns:

```
<console>:21: error: inferred type arguments [Int,Nothing,Int,Array[Int],Nothing] 
do not conform to method statefulMap's type parameter bounds 
[A,B,S,Repr <: scala.collection.immutable.Seq[A],That]

statefulMap(aCycleModulo4, 0)(addAndIncrement)
^
```

I chose `sc.SeqLike` and `sci.Seq` over `TraversableLike` and `Traversable` or `GenTraversableLike` and `GenTraversable` for a few reasons.  The first is that there are no iteration order guarantees for `Traversable`s so the function wouldn't be guaranteed to be idempotent (which I didn't like), whereas sequences do.  The reason I chose `Repr <: sci.Seq[A]` is that I like the idea that of immutability.  A third reason to choose immutable is that it precludes arrays.  If we want arrays, there should be a separate method because we would have to do an ***O***(***N***) array copy to get the tail and another ***O***(***N***) copy to produce the result.  This is both slow in terms of runtime and excessive in terms of memory usage.